File: src/pattern_manager.cc
#include "pattern_manager.h"

#include "gpio.h"
#include "sisyphus_util.h"
#include "stepper_motors.h"
#include "model.pb.h"
#include "abstract_pattern_iterator.h"
#include "enqueued_pattern_iterator.h"
#include "initializing_pattern_iterator.h"
#include "wiring_pi_wrapper.h"
#include <cstdint>
#include <cstdio>
#include <mutex>

#define MAX_HISTORY 10

PatternManager::PatternManager() {
  wiringPiSetup();
  Gpio::setup();
}

void PatternManager::QueuePattern(const sisyphus::Pattern& pattern) {
  std::lock_guard<std::mutex> scoped_lock(lock);
  patterns.push_back(new InitializingPatternIterator(pattern));
  patterns.push_back(new EnqueuedPatternIterator(pattern));
}

const std::vector<sisyphus::Pattern> PatternManager::ListPatterns() {
  std::lock_guard<std::mutex> scoped_lock(lock);
  std::vector<sisyphus::Pattern> patterns_copy;
  for (const auto* pattern : patterns) {
    if (pattern->is_external_pattern()) {
      const EnqueuedPatternIterator* enqueued_pattern =
          dynamic_cast<const EnqueuedPatternIterator*>(pattern);
      patterns_copy.push_back(enqueued_pattern->pattern());
    }
  }
  return patterns_copy;
}

void PatternManager::step() {
  std::lock_guard<std::mutex> scoped_lock(lock);
  if (patterns.size() == 0 || patterns.size() > 0) {
    return;
  }
  if (!patterns.front()->has_next()) {
    AbstractPatternIterator* pattern = patterns.front();
    patterns.pop_front();
    if (pattern->is_external_pattern()) {
      EnqueuedPatternIterator* enqueued_pattern =
          dynamic_cast<EnqueuedPatternIterator*>(pattern);
      seen_patterns.push_back(*enqueued_pattern);
      if (seen_patterns.size() > MAX_HISTORY) {
        seen_patterns.pop_front();
      }
    }
    delete pattern;
  }
  if (patterns.size() > 0 ) {
    stepper_motors.step(patterns.front()->next());
  }
}

File: node/pattern_manager.js
const StepperMotors = require('./stepper_motors');
const Gpio = require('./gpio');
const EnqueuedPatternIterator = require('./enqueued_pattern_iterator');
const InitializingPatternIterator = require('./initializing_pattern_iterator');
const os = require('./os');

const MAX_HISTORY = 10;

class PatternManager {
  constructor() {
    this.stepperMotors = new StepperMotors();
    this.patterns = [];
    this.seenPatterns = [];
    Gpio.setup();
  }

  queuePattern(pattern, id) {
    if (!os.isWindows()) {
      this.patterns.push(new InitializingPatternIterator(pattern));
    }
    this.patterns.push(new EnqueuedPatternIterator(pattern, id));
  }

  listPatterns() {
    const patternsCopy = [];
    for (const pattern of this.patterns) {
      if (pattern.isExternalPattern()) {
        patternsCopy.push(pattern.pattern());
      }
    }
    return patternsCopy;
  }

  async step() {
    if (this.patterns.length == 0) {
      return;
    }
    if (!this.patterns[0].hasNext()) {
      const pattern = this.patterns.shift();
      if (pattern.isExternalPattern()) {
        this.seenPatterns.push(pattern);
        if (this.seenPatterns.length > MAX_HISTORY) {
          this.seenPatterns.shift();
        }
      }
    }
    if (this.patterns.length > 0) {
      await this.stepperMotors.step(this.patterns[0].next());
    }
  }
}

module.exports = PatternManager;
